README.md
# node-graylog2
[![NPM version](http://img.shields.io/npm/v/graylog2.svg?style=flat-square)](https://www.npmjs.org/package/graylog2) [![NPM license](http://img.shields.io/npm/l/graylog2.svg?style=flat-square)](https://www.npmjs.org/package/graylog2)

Graylog2 client library for Node.js, based on node-graylog. This
has been heavily modified to the point where there is not much left
of the original; however, this library should still be compatible
with the old one, except for configuration and the GLOBAL function setup
(some optional arguments in logging calls are not supported; they will be
logged as additional data).

** New: ** Chunked [GELF](https://github.com/Graylog2/graylog2-docs/wiki/GELF)
is now supported.

## Synopsis

### Available functions

* graylog.emergency
* graylog.alert
* graylog.critical
* graylog.error
* graylog.warning
* graylog.notice
* graylog.info
* graylog.debug

### Code snippets

```javascript
var graylog2 = require("graylog2");
var logger = new graylog2.graylog({
    servers: [
        { 'host': 127.0.0.1, port: 12201 },
        { 'host': 127.0.0.2, port: 12201 }
    ],
    hostname: 'server.name', // the name of this host
                             // (optional, default: os.hostname())
    facility: 'Node.js',     // the facility for these log messages
                             // (optional, default: "Node.js")
    bufferSize: 1350         // max UDP packet size, should never exceed the
                             // MTU of your system (optional, default: 1400)
});

logger.on('error', function (error) {
    console.error('Error while trying to write to graylog2:', error);
});

```

Short message:

```javascript
logger.log("What we've got here is...failure to communicate");
```

Long message:

```javascript
logger.log("What we've got here is...failure to communicate", "Some men you just
    can't reach. So you get what we had here last week, which is the way he wants
    it... well, he gets it. I don't like it any more than you men.");
```

Short with additional data:

```javascript
logger.log("What we've got here is...failure to communicate", { cool: 'beans' });
```

Long with additional data:

```javascript
logger.log("What we've got here is...failure to communicate", "Some men you just
    can't reach. So you get what we had here last week, which is the way he wants
    it... well, he gets it. I don't like it any more than you men.",
    {
        cool: "beans"
    }
);
```

Flush all log messages and close down:
```javascript
logger.close(function(){
    console.log('All done - cookie now?');
    process.exit();
});
```

## Example

See `test.js`.

## What is graylog2 after all?

It's a miracle. Get it at http://www.graylog2.org/

## Installation

    npm install graylog2


CHANGELOG
0.2.0: September 20, 2017
* Bumped the GELF spec version to 1.1.

0.1.3: July 16, 2014
* Added asynchronous close(cb) function that closes and returns when all messages in the queue have been sent.

0.1.1: October 29, 2013
* The Emergency level was yielding "undefined", making Graylog2 fall back to the default Alert level.
* According to documentation log method should handle following invocation:
  `client.log('short message', {param: "value"})`, but instead of sending additional fields, JSON
  was treated like full message field.

0.1.0: September 3, 2013
* Replaced all console.error calls with emit('error').
  *Warning*: you really must listen for these or node will throw uncatchable exceptions!
* Bugfix: chunk count calculation was ignoring the header-overhead.
* All log operations now use a single udp socket.
* To be more friendly to lower MTUs, lowered the default buffer size to 1400 bytes.
* Added destroy() function that closes the UDP socket.
* Removed async requirement.
* Removed some unused variables.
* Removed the larger test files (256 and 512KB) because they don't fit the new default buffer limit.
* Removed unused test data.

0.0.12: March 12, 2012
* Dummy version bump in order to push to npm

0.0.11: March 12, 2012
* Removed dependency on node-compress-buffer
* Code cleanup
* Contributors mentioned

0.0.10: November 6, 2011
* Support node 0.6.0+

0.0.9: September 19, 2011
* Work around node 0.5.x crashes
* Added 0.5.x notes
* Added an important requirement of compress-buffer >= 0.4.0
* Added sayHello.js

0.0.8: July 12, 2011
* Minor console log output issue

0.0.7: June 23, 2011
* So much better console output

0.0.6: June 23, 2011
* GLOBAL.graylogSequence

0.0.5: May 14, 2011
* GLOBAL.graylogFacility
* Limit message size to 8k
* Misc typo fixes

0.0.2, 0.0.3, 0.0.4: May 12, 2011
* Exceptions in case the graylog2 is not running
* GLOBAL.graylogToConsole added
* CHANGELOG added
* Bugfixes

0.0.1: May 11, 2011
* initial release


test.js
var graylog = require('./graylog'),
    fs      = require('fs'),
    file,
    data,
    servers = [
        { 'host': '127.0.0.1', 'port': 12201 }
    ];


var client = new graylog.graylog({
        servers: servers,
        facility: 'Test logger / Node.JS Test Script'
    });

console.log('---------------------------------------------');
console.log('Sending three test as info, warning and error');
console.log('---------------------------------------------');
client.log('test1', 'i get this1', {cool: 'beans'});
client.warn('test2', 'i get this2', {cool: 'beans'});
client.error('test3', 'i get this3', {cool: 'beans'});
client.error('customTime', 'i get this3', {cool: 'beans'}, new Date('2012-10-10 13:20:31.619Z'));
console.log('');

console.log('---------------------------------------------');
console.log('Sending Sean Connery\' picture (as critical)');
console.log('---------------------------------------------');
file = './data/sean.jpg';
data = fs.readFileSync(file);
client.critical('My Nice Sean Connery Picture', data.toString(), {name: 'James Bond'});
console.log('');

console.log('---------------------------------------------');
console.log('Sending data of different sizes (as critical)');
console.log('---------------------------------------------');
for (var i = 4; i <= 128; i *= 2) {
    file = './data/' + i + '.dat';
    data = fs.readFileSync(file);
    console.log('sending', file);
    client.critical('Test 4 ' + file, data.toString(), {datafile: i + '.dat'});
}
console.log('');

console.log('---------------------------------------------');
console.log('Sending different parameters');
console.log('---------------------------------------------');
client.log('ParametersTest - Only short message');
client.log('ParametersTest - Short message and json', {cool: 'beans'});
client.log('ParametersTest - Short message and full message', 'Full message');
client.log('ParametersTest - Short Message with full message and json', 'Full message', {cool: 'beans'});
console.log('');

console.log('---------------------------------------------');
console.log('Sending without deflate');
console.log('---------------------------------------------');
client.deflate = 'never';
for (var i = 4; i <= 64; i *= 2) {
    file = './data/' + i + '.dat';
    data = fs.readFileSync(file);
    console.log('sending', file);
    client.critical('Test 4 ' + file, data.toString(), {datafile: i + '.dat'});
}
client.deflate = 'optimal';
console.log('');

client.close(function () {
    console.log('Insertion complete. Please check', 'http://' + servers[0].host + ':3000', 'and verify that insertion was successfull');
    console.log('');
});


graylog.js
var zlib         = require('zlib'),
    crypto       = require('crypto'),
    dgram        = require('dgram'),
    util         = require('util'),
    EventEmitter = require('events').EventEmitter,
    assert       = require('assert');

/**
 * Graylog instances emit errors. That means you really really should listen for them,
 * or accept uncaught exceptions (node throws if you don't listen for "error").
 */

var graylog = function graylog(config) {
    EventEmitter.call(this);

    this.config       = config;

    this.servers      = config.servers;
    this.client       = null;
    this.hostname     = config.hostname || require('os').hostname();
    this.facility     = config.facility || 'Node.js';
    this.deflate      = config.deflate || 'optimal';
    assert(
      this.deflate === 'optimal' || this.deflate === 'always' || this.deflate === 'never',
      'deflate must be one of "optimal", "always", or "never". was "' + this.deflate + '"');

    this._unsentMessages = 0;
    this._unsentChunks = 0;
    this._callCount   = 0;

    this._onClose = null;
    this._isDestroyed = false;

    this._bufferSize  = config.bufferSize || this.DEFAULT_BUFFERSIZE;
};

util.inherits(graylog, EventEmitter);

graylog.prototype.DEFAULT_BUFFERSIZE = 1400;  // a bit less than a typical MTU of 1500 to be on the safe side

graylog.prototype.level = {
    EMERG: 0,    // system is unusable
    ALERT: 1,    // action must be taken immediately
    CRIT: 2,     // critical conditions
    ERR: 3,      // error conditions
    ERROR: 3,    // because people WILL typo
    WARNING: 4,  // warning conditions
    NOTICE: 5,   // normal, but significant, condition
    INFO: 6,     // informational message
    DEBUG: 7     // debug level message
};

graylog.prototype.getServer = function () {
    return this.servers[this._callCount++ % this.servers.length];
};

graylog.prototype.getClient = function () {
    if (!this.client && !this._isDestroyed) {
        this.client = dgram.createSocket("udp4");

        var that = this;
        this.client.on('error', function (err) {
            that.emit('error', err);
        });
    }

    return this.client;
};

graylog.prototype.destroy = function () {
    if (this.client) {
        this.client.close();
        this.client.removeAllListeners();
        this.client = null;
        this._onClose = null;
        this._isDestroyed = true;
    }
};

graylog.prototype.emergency = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.EMERG);
};

graylog.prototype.alert = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.ALERT);
};

graylog.prototype.critical = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.CRIT);
};

graylog.prototype.error = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.ERROR);
};

graylog.prototype.warning = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.WARNING);
};
graylog.prototype.warn = graylog.prototype.warning;

graylog.prototype.notice = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.NOTICE);
};

graylog.prototype.info = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.INFO);
};
graylog.prototype.log = graylog.prototype.info;

graylog.prototype.debug = function (short_message, full_message, additionalFields, timestamp) {
    return this._log(short_message, full_message, additionalFields, timestamp, this.level.DEBUG);
};

graylog.prototype._log = function log(short_message, full_message, additionalFields, timestamp, level) {
    this._unsentMessages += 1;

    var payload,
        fileinfo,
        that    = this,
        field   = '',
        message = {
            version    : '1.0',
            timestamp  : (timestamp || new Date()).getTime() / 1000,
            host       : this.hostname,
            facility   : this.facility,
            level      : level
        };

    if (typeof(short_message) !== 'object' && typeof(full_message) === 'object' && additionalFields === undefined) {
        // Only short message and additional fields are available
        message.short_message   = short_message;
        message.full_message    = short_message;

        additionalFields = full_message;
    } else  if (typeof(short_message) !== 'object') {
        // We normally set the data
        message.short_message   = short_message;
        message.full_message    = full_message || short_message;
    } else if (short_message.stack && short_message.message) {

        // Short message is an Error message, we process accordingly
        message.short_message = short_message.message;
        message.full_message  = short_message.stack;

        // extract error file and line
        fileinfo = message.stack.split('\n')[0];
        fileinfo = fileinfo.substr(fileinfo.indexOf('('), fileinfo.indeOf(')'));
        fileinfo = fileinfo.split(':');

        message.file = fileinfo[0];
        message.line = fileinfo[1];

        additionalFields = full_message || additionalFields;
    } else {
        message.full_message = message.short_message = JSON.stringify(short_message);
    }

    // We insert additional fields
    for (field in additionalFields) {
        message['_' + field] = additionalFields[field];
    }

    // https://github.com/Graylog2/graylog2-docs/wiki/GELF
    if (message._id) {
        message.__id = message._id;
        delete message._id;
    }

    // Compression
    payload = new Buffer(JSON.stringify(message));

    function sendPayload(err, buffer) {
        if (err) {
            that._unsentMessages -= 1;
            return that.emitError(err);
        }

        // If it all fits, just send it
        if (buffer.length <= that._bufferSize) {
            that._unsentMessages -= 1;
            return that.send(buffer, that.getServer());
        }

        // It didn't fit, so prepare for a chunked stream

        var bufferSize = that._bufferSize;
        var dataSize   = bufferSize - 12;  // the data part of the buffer is the buffer size - header size
        var chunkCount = Math.ceil(buffer.length / dataSize);

        if (chunkCount > 128) {
            that._unsentMessages -= 1;
            return that.emitError('Cannot log messages bigger than ' + (dataSize * 128) +  ' bytes');
        }

        // Generate a random id in buffer format
        crypto.randomBytes(8, function (err, id) {
            if (err) {
                that._unsentMessages -= 1;
                return that.emitError(err);
            }

            // To be tested: what's faster, sending as we go or prebuffering?
            var server = that.getServer();
            var chunk = new Buffer(bufferSize);
            var chunkSequenceNumber = 0;

            // Prepare the header

            // Set up magic number (bytes 0 and 1)
            chunk[0] = 30;
            chunk[1] = 15;

            // Set the total number of chunks (byte 11)
            chunk[11] = chunkCount;

            // Set message id (bytes 2-9)
            id.copy(chunk, 2, 0, 8);

            function send(err) {
                if (err || chunkSequenceNumber >= chunkCount) {
                    // We have reached the end, or had an error (which will already have been emitted)
                    that._unsentMessages -= 1;
                    return;
                }

                // Set chunk sequence number (byte 10)
                chunk[10] = chunkSequenceNumber;

                // Copy data from full buffer into the chunk
                var start = chunkSequenceNumber * dataSize;
                var stop  = Math.min((chunkSequenceNumber + 1) * dataSize, buffer.length);

                buffer.copy(chunk, 12, start, stop);

                chunkSequenceNumber++;

                // Send the chunk
                that.send(chunk.slice(0, stop - start + 12), server, send);
            }

            send();
        });
    }

    if (this.deflate === 'never' || (this.deflate === 'optimal' && payload.length <= this._bufferSize)) {
      sendPayload(null, payload);
    } else {
      zlib.deflate(payload, sendPayload);
    }
};

graylog.prototype.send = function (chunk, server, cb) {
    var that = this,
        client = this.getClient();

    if (!client) {
        var error = new Error('Socket was already destroyed');

        this.emit('error', error);
        return cb(error);
    }

    this._unsentChunks += 1;

    client.send(chunk, 0, chunk.length, server.port, server.host, function (err) {
        that._unsentChunks -= 1;

        if (err) {
            that.emit('error', err);
        }

        if (cb) {
            cb(err);
        }

        if (that._unsentChunks === 0 && that._unsentMessages === 0 && that._onClose) {
            that._onClose();
        }
    });
};

graylog.prototype.emitError = function (err) {
    this.emit('error', err);

    if (this._unsentChunks === 0 && this._unsentMessages === 0 && this._onClose) {
        this._onClose();
    }
};

graylog.prototype.close = function (cb) {
    var that = this;

    if (this._onClose || this._isDestroyed) {
        return process.nextTick(function () {
            var error = new Error('Close was already called once');

            if (cb) {
                return cb(error);
            }

            that.emit('error', error);
        });
    }

    this._onClose = function () {
        that.destroy();

        if (cb) {
            cb();
        }
    };

    if (this._unsentChunks === 0 && this._unsentMessages === 0) {
        process.nextTick(function () {
            that._onClose();
        });
    }
};

exports.graylog = graylog;
